File: search/java/search/src/xyz/zrui/algorithm/search/BinarySearch.java
package xyz.zrui.algorithm.search;

public class BinarySearch {

    public static int rank(int key, int[] arr) {

        int hi = arr.length;
        int lo = 0;

        while (lo <= hi) {

            int mid = lo + (hi - lo) / 2;

            if (arr[mid] > key) {
                hi = mid - 1;
            } else if (arr[mid] < key) {
                lo = mid + 1;
            } else {
                return mid;
            }
        }
        return -1;

    }

    public static void main(String[] args) {
        int[] arr = {0,1,2,4,5,6,8,9};
        System.out.println(rank(6,arr));

    }




}

File: search/golang/search/binary_search.go
package search

import (
	"fmt"
)

// 递归实现
func BinaryRecSearch(arr []int, l, r, target int) int {

	mid := l + (r-l)/2
	if arr[mid] == target {
		return mid
	}
	if arr[mid] > target {
		return BinaryRecSearch(arr, l, mid-1, target)
	} else {
		return BinaryRecSearch(arr, mid+1, r, target)
	}

}

//二分查找法 返回索引值
func BinarySearch(arr []int, target int) int {

	l := 0
	r := len(arr) - 1

	for l <= r {

		mid := l + (r-l)/2

		if arr[mid] == target {
			fmt.Println(0)
			return mid
		}
		if arr[mid] < target {
			l = mid + 1
		} else {

			r = mid - 1
		}
	}
	return -1

}
